### PR TITLE
#764 Decrease the storage usage for Solana receipts 

### PR DESCRIPTION
**Pull request by otselnik**

### CODE REVIEW

File: proxy/indexer/indexer_base.py [hunk 1]
```diff
@@ -90,109 +91,108 @@ def run(self):
...
 
-                if sol_sign in [HISTORY_START, self._maximum_tx]:
-                    continue_flag = False
-                    break
-
-                tx_list.append((sol_sign, slot, tx_idx))
-                if len(tx_list) >= tx_per_request:
-                    poll_txs.append(tx_list)
+                if len(tx_list) >= INDEXER_POLL_COUNT:
+                    self.solana_signatures.add_signature(tx_list[0][0], tx_list[0][1])
```
Comment (afalaleev) [inline proxy/indexer/indexer_base.py:161]: Solana blocks can contain a lot of transactions.
If block contain more than INDEXER_POOL_COUNT transactions, the table with solana signatures will have more than 1 record with the same block slot.
You should set in this place flag and save checkpoint only on changing of block slot.

File: proxy/indexer/indexer_base.py [hunk 2]
```diff
@@ -90,109 +91,108 @@ def run(self):
...
-        )
 
-    def _get_signatures(self, before: Optional[str], limit: int) -> []:
-        response = self.solana.get_signatures_for_address(before, limit, FINALIZED)
+    def _get_signatures(self, before: Optional[str], until: Optional[str], limit: int) -> List[Dict[str, Union[int, str]]]:
+        opts: Dict[str, Union[int, str]] = {}
+        if before is not None:
+            opts["before"] = before
+        if until is not None:
+            opts["until"] = until
```
Comment (afalaleev) [inline proxy/indexer/indexer_base.py]: One month ago we fixed a bug with `until`.
Solana node returns two independent range of transactions in which the end of the first range was the begin of the second range, and at the same time the end of the second range was the begin of the first range.
For example:
```[tx1, tx2 ... tx3, tx4]  <-> [tx4, tx5 ... tx6, tx1]```

As a result, the indexer deadlooped on trying to get the signatures.
It is better to use only `before` and break on the getting `until`

File: proxy/indexer/indexer_base.py [hunk 3]
```diff
@@ -90,109 +91,108 @@ def run(self):
...
-            poll_txs.append(tx_list)
-        if len(poll_txs) > 0:
-            self._get_txs(poll_txs)
-
-        self.current_slot = current_slot
-        self.counter_ = 0
-        self._set_maximum_tx(maximum_tx)
+        signatures = []
+        for signature, _ in reversed(tx_list):
+            signatures.append(signature)
```
Comment (afalaleev) [inline proxy/indexer/indexer_base.py]: This is only comment: you can return tx_list and reverse after getting receipts

File: proxy/indexer/indexer_base.py [hunk 4]
```diff
@@ -90,109 +91,108 @@ def run(self):
...
-        tx_per_request = 20
+        for signature in signatures:
+            if signature not in self._tx_receipts:
+                self.error(f'{signature} receipt not found')
+                continue
 
+            tx = self._tx_receipts[signature]
+            slot = tx['slot']
+            if stop_slot and slot > stop_slot:
+                break
```
Comment (afalaleev) [inline proxy/indexer/indexer_base.py:120]: As I understand, on the next call `get_tx_receipts()` , `gather_unknown_transaction()` will return the same list of transactions and they will be passed to the receipt parser?

**Review by afalaleev — CHANGES_REQUESTED** [review]
Looks good! But several issues should be fixed.

File: proxy/indexer/indexer_base.py (hunk 2 above)
Comment (otselnik) [reply proxy/indexer/indexer_base.py]: fixed

File: proxy/indexer/indexer_base.py (hunk 4 above)
Comment (otselnik) [reply proxy/indexer/indexer_base.py:120]: `self._maximum_tx` is overwritten every cycle and `gather_unknown_transaction` will return jnly new transactions till `self._maximum_tx`

File: proxy/indexer/indexer_base.py (hunk 1 above)
Comment (otselnik) [reply proxy/indexer/indexer_base.py:161]: Fixed by making slot unique.

File: proxy/indexer/indexer_base.py (hunk 3 above)
Comment (otselnik) [reply proxy/indexer/indexer_base.py]: Optimized

File: proxy/indexer/indexer_base.py (hunk 4 above)
Comment (afalaleev) [reply proxy/indexer/indexer_base.py:120]: Ok.
We can set_maximum_tx in two cases:
- on the break
- on the end of loop

Currently, we have an surplus update on each call.

File: proxy/indexer/indexer_base.py (hunk 2 above)
Comment (afalaleev) [reply proxy/indexer/indexer_base.py]: I don't see the fix, the source code is the same

**Review by afalaleev — CHANGES_REQUESTED** [review]

File: proxy/indexer/indexer_base.py (hunk 2 above)
Comment (otselnik) [reply proxy/indexer/indexer_base.py]: https://github.com/neonlabsorg/proxy-model.py/blob/764_do_not_store_solana_receipts_in_indexer/proxy/indexer/indexer_base.py#L148

File: proxy/indexer/indexer_base.py (hunk 4 above)
Comment (otselnik) [reply proxy/indexer/indexer_base.py:120]: But on the break it not processed yet and must be processed on next call `get_tx_receipts`

Comment (otselnik) [reply proxy/indexer/indexer_base.py:120]: I update maximum_tx on each cycle for case when outer loop processing receipts catches error and throws. In this case if i would updated maximum_tx on end of loop `gather_unknown_transactions` will request same amount of signatures as previous call.

Comment (otselnik) [reply proxy/indexer/indexer_base.py:120]: I must agree that it is pointless, since when new transactions will come there will be no profit

**Review by afalaleev — APPROVED** [review]

**Pull request merged**